### PR TITLE
ddcci-plasmoid: init at 0.1.10

### PR DESCRIPTION
**Pull request by sund3RRR**
## Description of changes
[ddcci-plasmoid](https://github.com/davidhi7/ddcci-plasmoid) - KDE Plasma Widget for external monitor brightness adjustment using `ddcutil`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sund3RRR** [comment]
@SuperSandro2000

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, callPackage
+}:
+let
+  ddcci-plasmoid-backend = callPackage ./ddcci-plasmoid-backend { };
+in
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/default.nix:8]: We should move that to toplevel

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  ];
+
+  nativeCheckInputs = with python3Packages; [
+    pytest
+    tox
+    python-lsp-ruff
+  ];
+
+  postFixup = ''
+    wrapProgram $out/bin/ddcci_plasmoid_backend \
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix:32]: Is this double wrapping?

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  ];
+
+  propagatedBuildInputs = with python3Packages; [
+    fasteners
+  ];
+
+  nativeCheckInputs = with python3Packages; [
+    pytest
+    tox
+    python-lsp-ruff
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix:28]:
```suggestion
```
we are not interested in formatting and linting in checks

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    poetry-core
+  ];
+
+  propagatedBuildInputs = with python3Packages; [
+    fasteners
+  ];
+
+  nativeCheckInputs = with python3Packages; [
+    pytest
+    tox
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix:27]:
```suggestion
```
tox will most likely not work as expected in the nix sandbox

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  nativeBuildInputs = with python3Packages; [
+    poetry-core
+  ];
+
+  propagatedBuildInputs = with python3Packages; [
+    fasteners
+  ];
+
+  nativeCheckInputs = with python3Packages; [
+    pytest
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix:26]:
```suggestion
    pytestCheckHook
```

File: pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+, ddcutil
+}:
+python3Packages.buildPythonPackage rec {
+  pname = "ddcci_plasmoid_backend";
+  version = "0.1.10";
+
+  format = "pyproject";
+
+  src = fetchPypi {
+    inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/desktops/plasma-5/plasmoids/ddcci-plasmoid/ddcci-plasmoid-backend/default.nix:13]:
```suggestion
  pname = "ddcci-plasmoid-backend";
  version = "0.1.10";

  format = "pyproject";

  src = fetchPypi {
    pname = "ddcci_plasmoid_backend";
    inherit version;
```
see python language docs in nixpkgs

**Review by SuperSandro2000 — COMMENTED** [review]
I think we should place those as a normal package outside of plasma

/cc @K900

**Comment by sund3RRR** [comment]
@SuperSandro2000

**Comment by K900** [comment]
Do we really need this? You can override powerdevil to handle DDC/CI, or use the kernel module.

**Comment by sund3RRR** [comment]
> Do we really need this? You can override powerdevil to handle DDC/CI, or use the kernel module.

It seems to me that kernel module has undefined behavior at system boot (at least on my system it either loads or it doesn't 50/50).
```
ddcci 11-0037: core device [6e] probe failed: -19
```
`ddcutil` with plasmoid works well.

**Comment by K900** [comment]
You can apply this as an override to powerdevil if you're confident DDC/CI works on your hardware: https://github.com/NixOS/nixpkgs/commit/31b1d66bfe0a88c424ac93437bc570c4ac99bc9a

**Comment by sund3RRR** [comment]
> You can apply this as an override to powerdevil if you're confident DDC/CI works on your hardware: [31b1d66](https://github.com/NixOS/nixpkgs/commit/31b1d66bfe0a88c424ac93437bc570c4ac99bc9a)

You said in your [pull request](https://github.com/NixOS/nixpkgs/pull/172078) that the integration of ddcutil in powerdevil is really buggy and it is better not to use it. What is the point for me to rebuild the system package if I can just install a well-functioning extension?

**Comment by sund3RRR** [comment]
Damn, looks like I broke a branch

**Comment by K900** [comment]
You generally want to rebase your changes on top of master, not merge master into your branch.

**Comment by sund3RRR** [comment]
Is it possible to fix it?

**Comment by K900** [comment]
`git rebase -i origin/master` and remove all the commits except yours.

**Comment by sund3RRR** [comment]
> `git rebase -i origin/master` and remove all the commits except yours.

Thanks, man. I wanted to update the branch to switch from kf5 to kf6, because `ddcci-plasmoid` has a version for plasma 6. Returning to the discussion of whether we need this, I think yes, because the ddcci kernel module does not detect the monitor if it does not respond within 60ms at the time of boot, so there is an undefined behaviour and driver are not always loaded.
Overriding powerdevil is also not the best idea, as far as I understand.
The plasmoid works well and always, so I like it.
If we decide that we need it, then I will update the version and switch to kf6. If not, then I will close the PR.

**Comment by sund3RRR** [comment]
So, I built this on plasma 6 and it works great

**Comment by sund3RRR** [comment]
@SuperSandro2000 @K900

**Pull request closed without merging**